tests/metagen/typegraphs/sample/ts_upload/client.ts

```diff
@@ -0,0 +1,878 @@
+
+      if (argumentTypes) {
+        // make sure the arg is of the expected form
+        let arg = instanceSelection;
+        if (Array.isArray(arg)) {
+          arg = arg[0];
+        }
+        // TODO: consider bringing in Zod (after hoisting impl into common lib)
+        if (typeof arg != "object" || arg === null) {
+          throw new Error(
+            `node at ${parentPath}.${instanceName} is a node ` +
+              `that requires arguments object but detected argument ` +
+              `is typeof ${typeof arg}`,
+          );
+        }
+
+        const expectedArguments = new Map(Object.entries(argumentTypes));
+        node.args = {};
+        for (const [key, value] of Object.entries(arg)) {
+          const typeName = expectedArguments.get(key);
+          // TODO: consider logging a warning if `_` is detected incase user passes
+          // Selection as arg
+          if (!typeName) {
+            throw new Error(
+              `unexpected argument ${key} at ${parentPath}.${instanceName}`,
+            );
+          }
+          expectedArguments.delete(key);
+          node.args[key] = { typeName, value };
+        }
+      }
+
+      if (subNodes) {
+        // sanity check selection object
+        let subSelections = instanceSelection;
+        if (argumentTypes) {
+          if (!Array.isArray(subSelections)) {
+            throw new Error(
+              `node at ${parentPath}.${instanceName} ` +
+                `is a composite that takes an argument ` +
+                `but selection is typeof ${typeof subSelections}`,
+            );
+          }
+          subSelections = subSelections[1];
+        } else if (Array.isArray(subSelections)) {
+          throw new Error(
+            `node at ${parentPath}.${instanceName} ` +
+              `is a composite that takes no arguments ` +
+              `but selection is typeof ${typeof subSelections}`,
+          );
+        }
+        if (typeof subSelections != "object") {
+          throw new Error(
+            `node at ${parentPath}.${nodeName} ` +
+              `is a no argument composite but first element of ` +
+              `selection is typeof ${typeof nodeSelection}`,
+          );
+        }
+
+        node.subNodes = _selectionToNodeSet(
+          // assume it's a Selection. If it's an argument
+          // object, mismatch between the node desc should hopefully
+          // catch it
+          subSelections as Selection,
+          subNodes,
+          `${parentPath}.${instanceName}`,
+        );
+      }
+
+      out.push(node);
+    }
+  }
+  foundNodes.delete("_");
+  if (foundNodes.size > 0) {
+    throw new Error(
+      `unexpected nodes found in selection set at ${parentPath}: ${[
+        ...foundNodes,
+      ]}`,
+    );
+  }
+  return out;
+}
+
+/* Query node types section */
+
+type SelectNode<_Out = unknown> = {
+  nodeName: string;
+  instanceName: string;
+  args?: NodeArgs;
+  subNodes?: SelectNode[];
```

Comment:
I rec collecting the found files from the `NodeMeta`s here. This would happen in _selectionToNodeSet`. You'd definitely prefer this approach in the Rust implementation which is way more strict.

In fact, I recommend doing the rust impl first since it's restrictions will provide more guidance.